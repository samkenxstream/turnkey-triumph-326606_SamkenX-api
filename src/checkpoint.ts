import { Provider } from 'starknet';
import Promise from 'bluebird';
import mysql from './mysql';

export default class Checkpoint {
  public config;
  public action;
  public provider: Provider;

  constructor(config, action) {
    this.config = config;
    this.action = action;
    this.provider = new Provider({ network: this.config.network });
  }

  async getStartBlock() {
    let start = 0;
    const lastBlock = await mysql.queryAsync('SELECT * FROM checkpoint LIMIT 1');
    const nextBlock = lastBlock[0].number + 1;
    this.config.sources.forEach(source => {
      start = start === 0 || start > source.start ? source.start : start;
    });
    return nextBlock > start ? nextBlock : start;
  }

  async start() {
    console.log('Start');
    const nextBlock = await this.getStartBlock();
    return await this.next(nextBlock);
  }

  async next(blockNum: number) {
    let block: any;
    try {
      block = await this.provider.getBlock(blockNum);
    } catch (e) {
      console.log('Get block failed', blockNum, JSON.stringify(e).slice(0, 256));
      await Promise.delay(12e3);
      return this.next(blockNum);
    }
    await this.handleBlock(block);
    const query = 'UPDATE checkpoint SET number = ?';
    await mysql.queryAsync(query, [block.block_number]);
    return this.next(blockNum + 1);
  }

  async handleBlock(block) {
    console.log('Handle block', block.block_number);
    for (const receipt of block.transaction_receipts)
      await this.handleTx(block, block.transactions[receipt.transaction_index], receipt);
    // console.log('Handle block done', block.block_number);
  }

  async handleTx(block, tx, receipt) {
    // console.log('Handle tx', tx.transaction_index);
    for (const source of this.config.sources) {
      if (source.contract === tx.contract_address) await this.action(source, block, tx, receipt);
    }
    // console.log('Handle tx done', tx.transaction_index);
  }
}
