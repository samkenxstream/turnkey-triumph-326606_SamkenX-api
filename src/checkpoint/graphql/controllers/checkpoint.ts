import {
  GraphQLFieldConfig,
  GraphQLFieldConfigMap,
  GraphQLInt,
  GraphQLList,
  GraphQLNonNull,
  GraphQLString
} from 'graphql';
import {
  queryCheckpoints,
  queryLastIndexedBlock,
  queryLatestStarknetBlock,
  ResolverContext
} from '../resolvers';

/**
 * Controller for generating query and mutation schemas relating
 * to core checkpoints data.
 *
 */
export class GqlCheckpointController {
  /**
   * This generate graphql Query fields required to introspect
   * how Checkpoint is performing.
   *
   * Fields exposed are prefixed with an underscore, so they are
   * distinguishable from the Entity Queries.
   *
   */
  public generateQueryFields(): GraphQLFieldConfigMap<any, ResolverContext> {
    return {
      _last_indexed_block: this.lastIndexedBlockField(),
      _latest_starknet_block: this.latestStarknetBlockField(),
      _checkpoints: this.checkpointsField()
    };
  }

  private lastIndexedBlockField(): GraphQLFieldConfig<any, ResolverContext> {
    return {
      type: GraphQLInt,
      description: 'Fetch the last block Checkpoint has processed.',
      resolve: queryLastIndexedBlock
    };
  }

  private latestStarknetBlockField(): GraphQLFieldConfig<any, ResolverContext> {
    return {
      type: GraphQLInt,
      description: `Fetch the latest block number on Starknet.
        Useful for comparing how far behind this Checkpoint's _last_indexed_block is behind the latest block. 
      NOTE: Querying this field can be quite slow.`,
      resolve: queryLatestStarknetBlock
    };
  }

  private checkpointsField(): GraphQLFieldConfig<any, ResolverContext> {
    return {
      type: new GraphQLList(GraphQLInt),
      description: 'Fetch starknet blocks (checkpoints) where a contracts interaction exists.',
      args: {
        contract: {
          type: new GraphQLNonNull(GraphQLString),
          description: 'contract address to fetch checkpoint blocks'
        },
        fromBlock: {
          type: GraphQLInt,
          description: 'filter result starting from this block'
        },
        size: {
          type: GraphQLInt,
          description:
            'maximum size of blocks to return in results. defaults to 100. Max possible value is 1000.'
        }
      },
      resolve: queryCheckpoints
    };
  }
}
