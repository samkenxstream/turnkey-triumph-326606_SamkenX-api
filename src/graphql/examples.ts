export default `
query {
  votes {
    voter
    proposal
    choice
  }
}
`;
