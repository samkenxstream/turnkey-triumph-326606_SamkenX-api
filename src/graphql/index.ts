import { graphqlHTTP } from 'express-graphql';
import { makeExecutableSchema } from '@graphql-tools/schema';
import checkpoint from './checkpoint';
import proposals from './proposals';
import proposal from './proposal';
import votes from './votes';
import defaultQuery from './examples';
import typeDefs from './schema';

const rootValue = {
  Query: {
    checkpoint,
    proposals,
    proposal,
    votes
  }
};
const schema = makeExecutableSchema({ typeDefs, resolvers: rootValue });

export default graphqlHTTP({
  schema,
  rootValue,
  graphiql: { defaultQuery }
});
