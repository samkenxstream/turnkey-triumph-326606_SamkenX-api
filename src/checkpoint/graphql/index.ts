import { graphqlHTTP } from 'express-graphql';
import { makeExecutableSchema } from '@graphql-tools/schema';
import { toGql, toQuery } from './utils';

export default function get(types) {
  const rootValue = { Query: toQuery(types) };
  const typeDefs = toGql(types);
  const schema = makeExecutableSchema({ typeDefs, resolvers: rootValue });
  return graphqlHTTP({ schema, rootValue });
}
