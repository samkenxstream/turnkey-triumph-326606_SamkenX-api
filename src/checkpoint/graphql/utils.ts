import { introspectionFromSchema } from 'graphql';
import { makeExecutableSchema } from '@graphql-tools/schema';
import mysql from '../mysql';

async function queryMulti(parent, args, context, info) {
  const params: any = [];
  let whereSql = '';
  if (args.where) {
    Object.entries(args.where).map((w) => {
      whereSql += !whereSql ? `WHERE ${w[0]} = ? ` : ` AND ${w[0]} = ?`;
      params.push(w[1]);
    });
  }
  const first = args?.first || 1000;
  const skip = args?.skip || 0;
  const orderBy = 'created';
  const orderDirection = 'DESC';
  params.push(orderBy, orderDirection, skip, first);
  return await mysql.queryAsync(
    `SELECT * FROM ${info.fieldName} ${whereSql} ORDER BY ? ? LIMIT ?, ?`,
    params
  );
}

async function querySingle(parent, args, context, info) {
  const query = `SELECT * FROM ${info.fieldName}s WHERE id = ? LIMIT 1`;
  const [item] = await mysql.queryAsync(query, [args.id]);
  return item;
}

export function toSql(typeDefs) {
  const schema = makeExecutableSchema({
    typeDefs: `type Query { x: String }\n${typeDefs}`
  });
  let sql = 'DROP TABLE IF EXISTS checkpoint;';
  sql += '\nCREATE TABLE checkpoint (number BIGINT NOT NULL, PRIMARY KEY (number));';
  sql += '\nINSERT checkpoint SET number = 0;';
  introspectionFromSchema(schema).__schema.types.forEach((type) => {
    const clone = JSON.parse(JSON.stringify(type));
    if (
      clone.kind === 'OBJECT' &&
      clone.fields &&
      !clone.name.startsWith('__') &&
      clone.name !== 'Query'
    ) {
      sql += `\n\nDROP TABLE IF EXISTS ${clone.name.toLowerCase()}s;`;
      sql += `\nCREATE TABLE ${clone.name.toLowerCase()}s (`;
      let sqlIndexes = ``;
      clone.fields.forEach((field) => {
        const fieldType = field.type.name;
        let sqlType = 'VARCHAR(128)';
        if (fieldType === 'Int') sqlType = 'INT(128)';
        if (fieldType === 'String') sqlType = 'VARCHAR(128)';
        if (fieldType === 'Text') sqlType = 'TEXT';
        sql += `\n  ${field.name} ${sqlType} NOT NULL,`;
        if (fieldType !== 'Text') sqlIndexes += `,\n  INDEX ${field.name} (${field.name})`;
      });
      sql += `\n  PRIMARY KEY (id) ${sqlIndexes}\n);`;
    }
  });
  return sql;
}

export function toGql(typeDefs) {
  let where = '';
  let gql = 'type Query {';
  const schema = makeExecutableSchema({
    typeDefs: `type Query { x: String }\n${typeDefs}`
  });
  introspectionFromSchema(schema).__schema.types.forEach((type) => {
    const clone = JSON.parse(JSON.stringify(type));
    if (
      clone.kind === 'OBJECT' &&
      clone.fields &&
      !clone.name.startsWith('__') &&
      clone.name !== 'Query'
    ) {
      where += `\n\ninput Where${clone.name} {`;
      gql += `\n  ${clone.name.toLowerCase()}s(`;
      gql += `\n    first: Int`;
      gql += `\n    skip: Int`;
      gql += `\n    orderBy: String`;
      gql += `\n    orderDirection: String`;
      gql += `\n    where: Where${clone.name}`;
      gql += `\n  ): [${clone.name}]`;
      gql += `\n  ${clone.name.toLowerCase()}(id: String): ${clone.name}`;
      clone.fields.forEach((field) => {
        const fieldType = field.type.name;
        if (fieldType !== 'Text') {
          where += `\n  ${field.name}: ${fieldType}`;
          where += `\n  ${field.name}_in: [${fieldType}]`;
          if (fieldType === 'Int') {
            where += `\n  ${field.name}_gt: ${fieldType}`;
            where += `\n  ${field.name}_gte: ${fieldType}`;
            where += `\n  ${field.name}_lt: ${fieldType}`;
            where += `\n  ${field.name}_lte: ${fieldType}`;
          }
        }
      });
      where += `\n}`;
    }
  });
  gql += `\n}\n${where}\n\n${typeDefs}`;
  return gql;
}

export function toQuery(typeDefs) {
  const query = {};
  const schema = makeExecutableSchema({
    typeDefs: `type Query { x: String }\n${typeDefs}`
  });
  introspectionFromSchema(schema).__schema.types.forEach((type) => {
    const clone = JSON.parse(JSON.stringify(type));
    if (
      clone.kind === 'OBJECT' &&
      clone.fields &&
      !clone.name.startsWith('__') &&
      clone.name !== 'Query'
    ) {
      query[`${clone.name.toLowerCase()}s`] = queryMulti;
      query[clone.name.toLowerCase()] = querySingle;
    }
  });
  return query;
}
