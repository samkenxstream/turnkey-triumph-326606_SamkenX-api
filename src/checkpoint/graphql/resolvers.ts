import { Provider } from 'starknet';
import { AsyncMySqlPool } from '../mysql';
import { CheckpointsStore, MetadataId } from '../stores/checkpoints';
import { Logger } from '../utils/logger';

/**
 *
 */
export interface ResolverContext {
  log: Logger;
  mysql: AsyncMySqlPool;
  checkpointsStore: CheckpointsStore;
  starknetProvider: Provider;
}

export async function queryMulti(parent, args, context: ResolverContext, info) {
  const { log, mysql } = context;

  const params: any = [];
  let whereSql = '';
  if (args.where) {
    Object.entries(args.where).map(w => {
      whereSql += !whereSql ? `WHERE ${w[0]} = ? ` : ` AND ${w[0]} = ?`;
      params.push(w[1]);
    });
  }
  const first = args?.first || 1000;
  const skip = args?.skip || 0;
  const orderBy = 'created';
  const orderDirection = 'DESC';
  params.push(skip, first);

  const query = `SELECT * FROM ${info.fieldName} ${whereSql} ORDER BY ${orderBy} ${orderDirection} LIMIT ?, ?`;
  log.debug({ sql: query, args }, 'executing multi query');

  return await mysql.queryAsync(query, params);
}

export async function querySingle(parent, args, context: ResolverContext, info) {
  const { log, mysql } = context;

  const query = `SELECT * FROM ${info.fieldName}s WHERE id = ? LIMIT 1`;
  log.debug({ sql: query, args }, 'executing single query');

  const [item] = await mysql.queryAsync(query, [args.id]);
  return item;
}

export async function queryLastIndexedBlock(_parent, _args, ctx: ResolverContext) {
  return ctx.checkpointsStore.getMetadataNumber(MetadataId.LastIndexedBlock);
}

// returns the latest starknet block visible to this node.
export async function queryLatestStarknetBlock(_parent, _args, ctx: ResolverContext) {
  // this might need some rate limiting to avoid being abused
  // because it is quite slow.
  const latestBlock = await ctx.starknetProvider.getBlock();
  return latestBlock.block_number;
}

const MaxCheckpointQuerySize = 1000;
export async function queryCheckpoints(_parent, args, ctx: ResolverContext) {
  const fromBlock = args.fromBlock || 0;
  const contracts = [args.contract];
  const size = args.size || 100;

  if (size > MaxCheckpointQuerySize) {
    throw new Error(`query 'size' exceeds max value. max value is ${MaxCheckpointQuerySize}`);
  }

  const checkpoints = await ctx.checkpointsStore.getNextCheckpointBlocks(
    fromBlock,
    contracts,
    size
  );
  return checkpoints;
}
