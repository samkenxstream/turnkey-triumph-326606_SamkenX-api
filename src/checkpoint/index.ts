import { Provider } from 'starknet';
import { starknetKeccak } from 'starknet/utils/hash';
import Promise from 'bluebird';
import mysql from './mysql';
import { toSql } from './graphql/utils';
import getGraphQL from './graphql';

export default class Checkpoint {
  public config;
  public writer;
  public schema;
  public graphql;
  public provider: Provider;

  constructor(config, writer, schema) {
    this.config = config;
    this.writer = writer;
    this.schema = schema;
    this.graphql = getGraphQL(schema);
    this.provider = new Provider({ network: this.config.network });
  }

  async getStartBlock() {
    let start = 0;
    const lastBlock = await mysql.queryAsync('SELECT * FROM checkpoint LIMIT 1');
    const nextBlock = lastBlock[0].number + 1;
    this.config.sources.forEach(source => {
      start = start === 0 || start > source.start ? source.start : start;
    });
    return nextBlock > start ? nextBlock : start;
  }

  async start() {
    console.log('Start');
    const nextBlock = await this.getStartBlock();
    return await this.next(nextBlock);
  }

  async next(blockNum: number) {
    let block: any;
    try {
      block = await this.provider.getBlock(blockNum);
    } catch (e) {
      console.log('Get block failed', blockNum, JSON.stringify(e).slice(0, 256));
      await Promise.delay(12e3);
      return this.next(blockNum);
    }
    await this.handleBlock(block);
    const query = 'UPDATE checkpoint SET number = ?';
    await mysql.queryAsync(query, [block.block_number]);
    return this.next(blockNum + 1);
  }

  async handleBlock(block) {
    console.log('Handle block', block.block_number);
    for (const receipt of block.transaction_receipts)
      await this.handleTx(block, block.transactions[receipt.transaction_index], receipt);
    // console.log('Handle block done', block.block_number);
  }

  async handleTx(block, tx, receipt) {
    // console.log('Handle tx', tx.transaction_index);
    for (const source of this.config.sources) {
      let contract = source.contract;
      contract = contract.length === 66 ? `0x${contract.slice(3)}` : contract;
      if (contract === tx.contract_address) await this.action(source, block, tx, receipt);
    }
    // console.log('Handle tx done', tx.transaction_index);
  }

  async action(source, block, tx, receipt) {
    console.log('Action for', source.contract, tx.type);
    console.log('Events', receipt.events.length);
    if (tx.type === 'DEPLOY' && source.deploy_fn)
      await this.writer[source.deploy_fn]({ source, block, tx, receipt });
    for (const sourceEvent of source.events) {
      for (const event of receipt.events) {
        if (starknetKeccak(sourceEvent.name).toString() === event.keys[0])
          await this.writer[sourceEvent.fn]({ source, block, tx, receipt });
      }
    }
  }

  async reset() {
    console.log('Reset');
    await mysql.queryAsync(toSql(this.schema));
  }
}
