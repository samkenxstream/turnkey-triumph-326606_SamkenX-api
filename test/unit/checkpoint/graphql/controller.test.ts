import { GraphQLObjectType, GraphQLSchema, printSchema } from 'graphql';
import { mock } from 'jest-mock-extended';
import { GqlCheckpointController } from '../../../../src/checkpoint/graphql/controllers/checkpoint';
import { GqlEntityController } from '../../../../src/checkpoint/graphql/controllers/entity';
import { AsyncMySqlPool } from '../../../../src/checkpoint/mysql';

describe('GqlEntityController', () => {
  describe('generateQueryFields', () => {
    it('should work', () => {
      const controller = new GqlEntityController(`
type Vote {
  id: Int!
  name: String
}
  `);
      const queryFields = controller.generateQueryFields();
      const querySchema = new GraphQLObjectType({
        name: 'Query',
        fields: queryFields
      });

      const schema = printSchema(new GraphQLSchema({ query: querySchema }));
      expect(schema).toMatchSnapshot();
    });

    // list of error table tests
    describe.each([
      {
        reason: 'non null object id',
        schema: `type Vote { id: String }`
      },
      {
        reason: 'object id is not scalar type',
        schema: `type Vote { id: Participant! }\n\n type Participant { id: Int! }`
      },
      {
        reason: 'object id is not scalar type 2',
        schema: `type Participant { id: [Int]! }`
      }
    ])('should fail for $reason', ({ schema }) => {
      const controller = new GqlEntityController(schema);
      expect(() => controller.generateQueryFields()).toThrowErrorMatchingSnapshot();
    });
  });

  describe('createEntityStores', () => {
    it('should work', async () => {
      const mockMysql = mock<AsyncMySqlPool>();
      const controller = new GqlEntityController(`
type Vote {
  id: Int!
  name: String
}
  `);
      await controller.createEntityStores(mockMysql);

      expect(mockMysql.queryAsync).toMatchSnapshot();
    });
  });
});

describe('GqlCheckpointController', () => {
  describe('generateQueryFields', () => {
    it('should work', () => {
      const controller = new GqlCheckpointController();
      const queryFields = controller.generateQueryFields();
      const querySchema = new GraphQLObjectType({
        name: 'Query',
        fields: queryFields
      });
      const schema = printSchema(new GraphQLSchema({ query: querySchema }));

      expect(schema).toMatchSnapshot();
    });

    it('should prefix all fields with underscore', () => {
      const controller = new GqlCheckpointController();
      const queryFields = controller.generateQueryFields();

      Object.keys(queryFields).forEach(fieldName => {
        expect(fieldName.substring(0, 1)).toEqual('_');
      });
    });
  });
});
