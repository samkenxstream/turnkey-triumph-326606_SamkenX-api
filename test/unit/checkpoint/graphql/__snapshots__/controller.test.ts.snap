// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `"'id' field for type Vote must be non nullable."`;

exports[` 2`] = `"'id' field for type Vote is not a scalar type."`;

exports[` 3`] = `"'id' field for type Participant is not a scalar type."`;

exports[`GqlCheckpointController generateQueryFields should work 1`] = `
"type Query {
  \\"\\"\\"Fetch the last block Checkpoint has processed.\\"\\"\\"
  _last_indexed_block: Int

  \\"\\"\\"
  Fetch the latest block number on Starknet.
          Useful for comparing how far behind this Checkpoint's _last_indexed_block is behind the latest block. 
        NOTE: Querying this field can be quite slow.
  \\"\\"\\"
  _latest_starknet_block: Int

  \\"\\"\\"
  Fetch starknet blocks (checkpoints) where a contracts interaction exists.
  \\"\\"\\"
  _checkpoints(
    \\"\\"\\"contract address to fetch checkpoint blocks\\"\\"\\"
    contract: String!

    \\"\\"\\"filter result starting from this block\\"\\"\\"
    fromBlock: Int

    \\"\\"\\"
    maximum size of blocks to return in results. defaults to 100. Max possible value is 1000.
    \\"\\"\\"
    size: Int
  ): [Int]
}"
`;

exports[`GqlEntityController createEntityStores should work 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "

DROP TABLE IF EXISTS votes;
CREATE TABLE votes (
  id INT(128) NOT NULL,
  name VARCHAR(128),
  PRIMARY KEY (id) ,
  INDEX id (id),
  INDEX name (name)
);",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`GqlEntityController generateQueryFields should work 1`] = `
"type Query {
  vote(id: Int!): Vote
  votes(first: Int, skip: Int, orderBy: String, orderDirection: String, where: WhereVote): [Vote]
}

type Vote {
  id: Int!
  name: String
}

input WhereVote {
  id_gt: Int
  id_gte: Int
  id_lt: Int
  id_lte: Int
  id: Int
  id_in: [Int]
  name: String
  name_in: [String]
}"
`;
